### PR TITLE
CRM-21111 Refactor include/exclude activity types into it's own function and im…

### PR DESCRIPTION
**Pull request by mattwire**
…prove performance (don't retrieve excluded activity types)

Before
----------------------------------------
API retrieved all activities, excluded activity types were removed afterwards.

After
----------------------------------------
API doesn't retrieve excluded activity types.

Technical Details
----------------------------------------
Extract activity type filtering to a new function filterActivityTypes();

Comments
----------------------------------------
The getActivities function is not actually in use anywhere yet so should be a relatively safe merge.

### CODE REVIEW

**Comment by mattwire** [comment]
test this please

File: CRM/Activity/BAO/Activity.php [hunk 1]
```diff
@@ -711,37 +708,11 @@ public static function getActivities($params, $getCount = FALSE) {
...
-        }
-        $activityParams['activity_type_id'] = array('IN' => $params['activity_type_id']);
-      }
-      else {
-        $activityParams['activity_type_id'] = CRM_Utils_Type::escape($params['activity_type_id'], 'Positive');
-      }
-    }
-    elseif (!empty($activityTypes) && count($activityTypes)) {
-      $activityParams['activity_type_id'] = array('IN' => array_keys($activityTypes));
+    if ($params['context'] != 'activity') {
```
Comment (eileenmcnaughton) [inline CRM/Activity/BAO/Activity.php]: not sure where $params['context'] came in from - original line is
```
-    if (!empty($params['activity_status_id'])) {
-      $activityParams['activity_status_id'] = array('IN' => explode(',', $params['activity_status_id']));
-    }
```

File: CRM/Activity/BAO/Activity.php [hunk 2]
```diff
@@ -840,10 +808,10 @@ public static function getActivities($params, $getCount = FALSE) {
         else {
           $activities[$id][$expectedName] = CRM_Utils_Array::value($apiKey, $activity);
           if ($apiKey == 'activity_type_id') {
-            $activities[$id]['activity_type'] = CRM_Utils_Array::value($activities[$id][$expectedName], $activityTypes);
+            $activities[$id]['activity_type'] = CRM_Core_PseudoConstant::getName(__CLASS__, 'activity_type_id', $activities[$id][$expectedName]);
```
Comment (eileenmcnaughton) [inline CRM/Activity/BAO/Activity.php]: I think it's better to put in the class name in case the function ever gets moved

File: CRM/Activity/BAO/Activity.php [hunk 3]
```diff
@@ -840,10 +808,10 @@ public static function getActivities($params, $getCount = FALSE) {
...
           }
           elseif ($apiKey == 'campaign_id') {
-            $activities[$id]['campaign'] = CRM_Utils_Array::value($activities[$id][$expectedName], $allCampaigns);
```
Comment (eileenmcnaughton) [inline CRM/Activity/BAO/Activity.php]: I actually feel a bit nervous about this line - as an unrelated change I feel like I have to work quite hard to validate it or let it skate through (& the latter is not good)

File: CRM/Activity/BAO/Activity.php [hunk 4]
```diff
@@ -859,6 +827,60 @@ public static function getActivities($params, $getCount = FALSE) {
...
+    if (empty($params['activity_type_id'])) {
+      $activityTypes = CRM_Activity_BAO_Activity::buildOptions('activity_type_id', 'get');
+    }
+
+    // If no activity types are specified or excluded, return the list of all active ones
+    if (empty($params['activity_type_id']) && empty($params['activity_type_exclude_id'])) {
+      if (!empty($activityTypes) && count($activityTypes)) {
+        return array('IN' => array_keys($activityTypes));
+      }
+      return NULL;
```
Comment (eileenmcnaughton) [inline CRM/Activity/BAO/Activity.php]: So this would only be hit if there were NO enabled activity types configured for the site at all? (in which case only inactive ones would be displayed)

File: CRM/Activity/BAO/Activity.php [hunk 5]
```diff
@@ -859,6 +827,60 @@ public static function getActivities($params, $getCount = FALSE) {
...
+    $activityTypes = array();
+
+    // If no activity types are specified, get all the active ones
+    if (empty($params['activity_type_id'])) {
+      $activityTypes = CRM_Activity_BAO_Activity::buildOptions('activity_type_id', 'get');
+    }
+
+    // If no activity types are specified or excluded, return the list of all active ones
+    if (empty($params['activity_type_id']) && empty($params['activity_type_exclude_id'])) {
+      if (!empty($activityTypes) && count($activityTypes)) {
```
Comment (eileenmcnaughton) [inline CRM/Activity/BAO/Activity.php]: I think !empty is enough without count

**Comment by eileenmcnaughton** [comment]
@mattwire I took a  look at this & I get what you are doing and I think it makes sense. I checked and this function has heavy test coverage - in fact it is currently ONLY reachable from the unit tests. The function has / had performance issues so we switched back to the deprecated function.

Given the above I'm happy to merge this if you can respond to the in-line comments. At least we will be in a better place if we later re-enable

File: CRM/Activity/BAO/Activity.php (hunk 4 above)
Comment (mattwire) [reply CRM/Activity/BAO/Activity.php]: Pretty unlikely to hit NULL yes.  As the return parameter is used as an API parameter it would return no activity types.

File: CRM/Activity/BAO/Activity.php (hunk 5 above)
Comment (mattwire) [reply CRM/Activity/BAO/Activity.php]: ok

File: CRM/Activity/BAO/Activity.php (hunk 3 above)
Comment (mattwire) [reply CRM/Activity/BAO/Activity.php]: Ok, changed back to original.

File: CRM/Activity/BAO/Activity.php (hunk 2 above)
Comment (mattwire) [reply CRM/Activity/BAO/Activity.php]: Ok, changed all instances

File: CRM/Activity/BAO/Activity.php (hunk 1 above)
Comment (mattwire) [reply CRM/Activity/BAO/Activity.php]: Now I'm not sure either.  I've changed it back to the original.

**Review by mattwire — COMMENTED** [review]
@eileenmcnaughton Let me know if you are happy with the changes and I'll squash the commits

**Comment by eileenmcnaughton** [comment]
@mattwire sure - please squash & cheer up jenkins re e-notice & we can merge
